src/features/broker-protection/actions/fill-form.js

```diff
@@ -47,35 +26,80 @@ export function fillForm (action, userData) {
+ * @param {HTMLElement} root
+ * @param {{selector: string; type: string}[]} elements
+ * @param {Record<string, any>} data
+ * @return {{
+ *     results: ({result: true} | {result: false; error: string})[],
+ *     warnings: {message:string}[]
+ * }}
+ */
+export function fillMany (root, elements, data) {
+    const results = []
+    const warnings = []
+
+    for (const element of elements) {
+        const inputElem = getElement(root, element.selector)
+        if (!inputElem) {
+            // todo: report this? it can occur when a selector is incorrect, but should we fail the entire action?
+            warnings.push({ message: `element not found for selector: "${element.selector}"` })
+            continue
+        }
+        if (element.type === '$file_id$') {
+            results.push(setImageUpload(inputElem))
+        } else if (element.type === '$generated_phone_number$') {
+            results.push(setValueForInput(inputElem, generatePhoneNumber()))
+        } else {
+            results.push(setValueForInput(inputElem, data[element.type]))
+        }
+    }
+    return { results, warnings }
+}
+
 /**
  * NOTE: This code comes from Autofill, the reasoning is to make React autofilling work on Chrome and Safari.
  *
  * Ensures the value is set properly and dispatches events to simulate real user action
  *
  * @param {HTMLElement} el
  * @param {string} val
- * @return {{result: boolean}}
+ * @return {{result: true} | {result: false; error: string}}
  */
 function setValueForInput (el, val) {
-    el.dispatchEvent(new Event('keydown', { bubbles: true }))
-
-    // Access the original setter (needed to bypass React's implementation on mobile)
-    // @ts-expect-error - Object will not be undefined on this case
-    const originalSet = Object.getOwnPropertyDescriptor(window.HTMLInputElement.prototype, 'value').set
-    originalSet?.call(el, val)
-
-    const events = [
-        new Event('input', { bubbles: true }),
-        new Event('keyup', { bubbles: true }),
-        new Event('change', { bubbles: true })
-    ]
-    events.forEach((ev) => el.dispatchEvent(ev))
-    // We call this again to make sure all forms are happy
-    originalSet?.call(el, val)
-    events.forEach((ev) => el.dispatchEvent(ev))
-    el.blur()
-
-    return { result: true }
+    // Access the original setters
+    // originally needed to bypass React's implementation on mobile
+    let target
+    if (el.tagName === 'INPUT') target = window.HTMLInputElement
+    if (el.tagName === 'SELECT') target = window.HTMLSelectElement
+
+    // Bail early if we cannot fill this element
+    if (!target) {
+        return { result: false, error: `input type was not supported: ${el.tagName}` }
+    }
+
+    try {
+        el.dispatchEvent(new Event('keydown', { bubbles: true }))
+
+        const originalSet = Object.getOwnPropertyDescriptor(target.prototype, 'value')?.set
+        originalSet?.call(el, val)
+
+        const events = [
```

Comment:
I noticed that the now the events we fire here for `select` are different than the [events fired for `select` in autofill](https://github.com/duckduckgo/duckduckgo-autofill/blob/6493e296934bf09277c03df45f11f4619711cb24/src/autofill-utils.js#L123-L136). Should we consider matching the implementation here?

Reply:
yes we should, good catch